### PR TITLE
Get branch name for PR in a better way

### PR DESCRIPTION
**Pull request by gautamabhishek46**
See [JENKINS-51867](https://issues.jenkins-ci.org/browse/JENKINS-51867)

### CODE REVIEW

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 1]
```diff
@@ -125,6 +138,36 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
+            refSpecs.set(i, refSpecs.get(i).trim());
+
+        return refSpecs;
+    }
+
+    private String getBranchName(String[] refSpecsArray){
+        boolean done = false;
+        StringBuilder branchName = new StringBuilder();
+        for(int i = 0; i < refSpecsArray.length && !done; i++){
+            if(refSpecsArray[i].equals("upstream") || refSpecsArray[i].equals("origin")){
```
Comment (gautamabhishek46) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: I saw bitbucket generated following two refspecs:
refs/heads/abhishekg1128/readmemd-edited-online-with-bitbucket-1529079381853:refs/remotes/origin/PR-1
refs/heads/master:refs/remotes/upstream/master

And GitHub generates
+refs/pull/3/head:refs/remotes/origin/PR-3 +refs/heads/master:refs/remotes/origin/master

File: src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java [hunk 2]
```diff
@@ -121,7 +121,7 @@ public String getPublishReportSnippet(ArrayList<String> paths){
         StringBuilder snippet = new StringBuilder();
 
         for(String p: paths)
-            snippet.append("junit '").append(p).append("'\n");
+            snippet.append("junit testResults: '").append(p).append(", allowEmptyResults: true").append("'\n");
```
Comment (grandvizier) [inline src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java]: these changes here seem unrelated to getting a branch name and therefor should be in a different PR 
(also note the commented code that shouldn't be merged...)

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 1 above)
Comment (grandvizier) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: this kind of info would be a great thing to add as a comment to the function definition

Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Sure I will add this on the top of the function.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 3]
```diff
@@ -125,6 +138,36 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
         return new CpsFlowExecution(script, false, owner);
     }
 
+    private List<String> getCleanRefspecs(List<UserRemoteConfig> userRemoteConfigs){
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: What this method does is not clear to me, in other words, what is the meaning of a "clean" refspec (is there such a thing as an unclean refspec?). It would be good to have a javadoc that describes what this method does. It should show an example of an input along with the output being produced.

Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: In refspecs there will be '+' symbol and some trailing white spaces, we need to remove them.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 4]
```diff
@@ -100,9 +102,20 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
         this.gitConfig = new GitConfig();
         this.gitConfig.setGitBranch(property.getBranch().getName());
 
+        listener.getLogger().println("Refspecs: ");
+
+        for(UserRemoteConfig urc: gitSCM.getUserRemoteConfigs()){
+            listener.getLogger().println(urc.getRefspec());
+        }
+
+        // FIXME This will fail if branch name will contain '/'
```
[the remark this answers is not in the document]
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: I think "/tree/" is only available in the URL when we browse the repository on GitHub only. I tried, but couldn't find a way to get this.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 3 above)
Comment (martinda) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Yes, I understand your answer, but you need to say what the method does in the javadoc because the code is not obvious in this case. The Twitter Java style guide has an [example of how to use javadoc for a method](https://github.com/twitter/commons/blob/master/src/java/com/twitter/common/styleguide.md#documenting-a-method).

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 5]
```diff
@@ -125,6 +138,36 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
+    private List<String> getCleanRefspecs(List<UserRemoteConfig> userRemoteConfigs){
+        List<String> refSpecs = new ArrayList<>();
+
+        for(UserRemoteConfig urc: userRemoteConfigs)
+            for(String s: urc.getRefspec().split("\\+", 0))
+                if(!s.equals(""))
+                    refSpecs.add(s);
+
+        for(int i = 0; i < refSpecs.size(); i++)
+            refSpecs.set(i, refSpecs.get(i).trim());
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: I feel the second loop is not needed if the first loop trims `s` before it is added.

Also, do not use single letter variable names, it makes it unsearchable (there are many hits on the letter s in the file). You could perhaps write `for (String refspec : urc.getRefSpec().split("\\+", 0).trim()) { ... }`.

Lastly, I do not know if Jenkins has coding style guidelines, but I suggest using curly braces even when they are optional.

File: src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java [hunk 6]
```diff
@@ -165,12 +165,12 @@ public String getStage(
...
-                                "credentialId: \"" + gitConfig.getCredentialsId() + "\"," +
-                                "url: \"" + gitConfig.getGitUrl() + "\"," +
-                                "branch: \"" + gitConfig.getGitBranch() + "\"" +
-                                "\n", 3);
+//                    if(gitConfig.getGitUrl() != null)
+//                        snippet += "\t\t\t" + addTabs("gitPush " +
+//                                "credentialId: \"" + gitConfig.getCredentialsId() + "\"," +
+//                                "url: \"" + gitConfig.getGitUrl() + "\"," +
+//                                "branch: \"" + gitConfig.getGitBranch() + "\"" +
+//                                "\n", 3);
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java]: @gautamabhishek46, every time you comment out code, you **MUST** say what future action is needed. Should this code be deleted? Should it be re-instated? Will it be replaced (and if so when and by who)?

If you comment out code because you don't need it, just delete it. If you keep it around "just in case", delete it: the previous commit has the code.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 5 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: In case of "+ someThing", the first loop will generate " someThing" hence there will be a space in the beginning. Hence second loop will be needed.

File: src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java (hunk 6 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java]: Comment: "Below commented code is to generate gitPush step, this is only commented for testing purpose." has been added.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 5 above)
Comment (grandvizier) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: I still think @martinda is right, the 2nd loop doesn't seem to be needed

currently your first for loop is like this:
```
String s: urc.getRefspec().split("\\+", 0)
```
and you say that `s = urc.getRefspec()` will give you "+ someThing"
but then you apply `s.split("\\+", 0)` to get " someThing"
and if you then apply `s.trim()` you'll get what you want with "someThing"

File: src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java (hunk 6 above)
Comment (kwhetstone) [reply src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java]: That works for me

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 5 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Ok, I have tested this thing, and I agree that the second loop in not needed. Hence it has been removed.

File: src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java (hunk 6 above)
Comment (martinda) [reply src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java]: I understand you comment it out for testing purposes, but you need to put a comment in the code stating that, for example:
```
// TODO Abhishek: code is commented out for testing purposes, it will be (removed|reinstated) later
```

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 7]
```diff
@@ -125,6 +131,48 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
         return new CpsFlowExecution(script, false, owner);
     }
 
+    // In refpecs there will be '+' symbol and some trailing white spaces, we need to remove them.
```
Comment (grandvizier) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: the comment is a good improvement, but it doesn't follow the javadoc style that would be preferred in this case

**Comment by gautamabhishek46** [comment]
@grandvizier does this look good ?

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 8]
```diff
@@ -101,8 +103,12 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
         this.gitConfig.setGitBranch(property.getBranch().getName());
 
         if(gitConfig.getGitBranch().startsWith("PR-")){
-            String[] refSpecs = gitSCM.getUserRemoteConfigs().get(0).getRefspec().split("/", 0);
-            gitConfig.setGitBranch(refSpecs[refSpecs.length - 1]);
+            for(String urc: getCleanRefSpecs(gitSCM.getUserRemoteConfigs())) {
```
Comment (grandvizier) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: It seems like it would make more sense to move this for loop logic into your `getBranchName` function, and then in there go through the refspecs and return the one you want to the `setGitBranch`

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 9]
```diff
@@ -146,23 +152,30 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
-
-    And GitHub generates following two similar refspecs in single string"
-    +refs/pull/3/head:refs/remotes/origin/PR-3 +refs/heads/master:refs/remotes/origin/master
-
-    */
-    private String getBranchName(String[] refSpecsArray){
+    /**
+     * Bit-bucket generates following two similar refSpecs for pull request:
+     * {@code refs/heads/abhishekg1128/readmemd-edited-online-with-bitbucket-1529079381853:refs/remotes/origin/PR-1}
+     * {@code refs/heads/master:refs/remotes/upstream/master}
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: There are two bitbucket implementations, 1) Bitbucket cloud (bitbucket.org) and 2) Bitbucket server. I guess these refs are for the Bitbucket cloud implementation?

On Bitbucket server, I see these references:
```
$ git ls-remote
From http://martin@localhost:7990/scm/spid/orange.git
f2a85eb1e8f6b47f94186d8e251c0bd984d04e4b	HEAD
bb1caf949f0300a82cd4242cc8b3d7469305a1a1	refs/heads/bugfix/exec-hello
f2a85eb1e8f6b47f94186d8e251c0bd984d04e4b	refs/heads/master
8fc7aafee8e7314ea98be9ea3f39a4f28306630e	refs/pull-requests/1/from
e30f42c437b9456b4f0f15e6f3be4690b3fdde8a	refs/pull-requests/1/merge
```

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 10]
```diff
@@ -146,23 +152,30 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
+     *
+     * And GitHub generates following two similar refspecs in single string"
+     * {@code +refs/pull/3/head:refs/remotes/origin/PR-3 +refs/heads/master:refs/remotes/origin/master}
+     * {@code getCleanRefSpecs} removes all the trailing spaces and unwanted '+' symbols from refSpecs.
+     *
+     * As branch name can contain {@code '/'}, this function will split the refSpec provided as parameter with
+     * {@code '/'} and look for {@code upstream} and {@code origin} keyword, when anyone of the two found it will
+     * assume that all the remaining elements correspond to the name of branch.
+     *
+     * @param refSpec One refSpec which will from which we need to get the name of branch.
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: How about "the refspec from which a branch name is extracted".

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 11]
```diff
@@ -146,23 +152,30 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
+     * @return List<String> List of strings containing refSpecs without trailing spaces and '+' symbols.
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: The method is called "getBranchName" so it should return the branch name. The `@return`description says that a list of strings is returned? The documentation does not match the code.

**Comment by martinda** [comment]
The javadoc are really helping me understand the intent. Thank you @gautamabhishek46 .

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 12]
```diff
@@ -146,23 +152,30 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
+    /**
+     * Bit-bucket generates following two similar refSpecs for pull request:
+     * {@code refs/heads/abhishekg1128/readmemd-edited-online-with-bitbucket-1529079381853:refs/remotes/origin/PR-1}
+     * {@code refs/heads/master:refs/remotes/upstream/master}
+     *
+     * And GitHub generates following two similar refspecs in single string"
+     * {@code +refs/pull/3/head:refs/remotes/origin/PR-3 +refs/heads/master:refs/remotes/origin/master}
+     * {@code getCleanRefSpecs} removes all the trailing spaces and unwanted '+' symbols from refSpecs.
+     *
+     * As branch name can contain {@code '/'}, this function will split the refSpec provided as parameter with
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Not critical but comments should be in the present tense as they are descriptive: "this function splits the refSpec...", and later "it assumes that all the remaining...".

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java [hunk 13]
```diff
@@ -146,23 +152,30 @@ public FlowExecution create(FlowExecutionOwner owner, TaskListener listener,
...
-    refs/heads/abhishekg1128/readmemd-edited-online-with-bitbucket-1529079381853:refs/remotes/origin/PR-1
-    refs/heads/master:refs/remotes/upstream/master
-
-    And GitHub generates following two similar refspecs in single string"
-    +refs/pull/3/head:refs/remotes/origin/PR-3 +refs/heads/master:refs/remotes/origin/master
-
-    */
-    private String getBranchName(String[] refSpecsArray){
+    /**
+     * Bit-bucket generates following two similar refSpecs for pull request:
```
Comment (martinda) [inline src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: Spelling: Bitbucket.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 9 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: References mentioned by you are shown in the terminal like this. RefSpecs is the mapping of remote references and local references. When we try to get refSpecs in the code it will be in the format that I mentioned in the comment for both bitbucket cloud and bitbucket server.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 11 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: It was a left after copy and paste I will correct it.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 8 above)
Comment (gautamabhishek46) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: I was thinking getBranchName as a function that will take a refSpec and returns the branch name. So the logic to get the refSpec from which we want to get the branch name should be out of this.

**Comment by gautamabhishek46** [comment]
Does this looks good  now.

File: src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java (hunk 8 above)
Comment (grandvizier) [reply src/main/java/io/jenkins/plugins/sprp/YAML_FlowDefinition.java]: yeah, I guess my comment is that in line 103 we set the branch name to the default, then we check if it starts with "PR-", and if it does, then we need to make use of the `getUserRemoteConfigs`. And so instead of looping through each "urc" and breaking on the first one that doesn't contain "PR-", I think it would make sense to call a function that does the looping and breaks on the first "valid" name. This way the logic of the function can have some clear unit tests, and if at some point there is an easier way to get the branch name it doesn't get mixed up with the for loop logic used here.
